File: legacy/src/scenes/tavern/player.ts
/**
 * The knight, in the tavern. Walk, face, stop at furniture. That's the whole job.
 *
 * This deliberately does NOT reuse `scenes/dungeon/entities/player.ts`. That
 * controller is ~1500 lines interleaved with pinball momentum, wall-launches,
 * rides, dodge-rolls, melee combo state and grid smashing, and it requires a
 * maze `Grid` to move at all. None of that exists here and none of it should.
 *
 * What IS reused is the art: the same cached `SpriteSheet` the dungeon builds
 * for the currently-held weapon, so the knight in the tavern is visibly the
 * same knight who just came up the stairs, holding the same thing.
 */
import * as THREE from "three";
import { state as dungeonState, activeWeapon } from "../dungeon/state";
import { buildSpriteSheet, createActorSprite } from "../dungeon/render/sprite";
import { makeKnightPaints } from "../dungeon/render/cel-painter";
import { Animator, facingFromVelocity, type Facing } from "../dungeon/render/animator";
import { PPU } from "../dungeon/constants";
import { screenDirToWorld } from "../dungeon/camera";
import type { InputHandle } from "../dungeon/input";
import { moveInRoom, SPAWN } from "./layout";
import { tavern, type TavernPlayer } from "./state";

/**
 * Walk speed, world units/sec.
 *
 * Was 3.4 — "slower than the dungeon, nothing is chasing you". That reasoning is
 * sound for the FICTION and wrong for the FEEL: the player arrives here straight
 * off the dungeon's PLAYER_SPEED 4.2, so a 19% drop reads as the controls having
 * gone soft, not as the room being calm. Worse, the tavern camera leans only
 * CAM_LEAN (0.5) toward the player (core.ts), which means the player's APPARENT
 * on-screen motion is exactly half their world motion — walk 1 unit, the frame
 * follows 0.5, so you slide 0.5 across the screen. The dungeon's dead-zone
 * camera holds still for small moves and gives you close to 1:1.
 *
 * So matching 4.2 exactly still reads slower than the dungeon. 4.6 is the
 * dungeon's speed plus a margin paid to the camera lean; crossing the 18-unit
 * room now takes ~3.9s instead of ~5.3s. Deliberately NOT taken further — this
 * is a hub with staged props, and past ~5 the knight starts skating.
 */
const WALK_SPEED = 4.6;
/** Hold shift to cross the room quickly without it feeling like a sprint mechanic. */
const HURRY_MULT = 1.5;
/**
 * Acceleration/deceleration, units/sec².
 *
 * Was 26 against the dungeon's MOVE_ACCEL 22. Raised with the speed so the RAMP
 * TIME doesn't grow: 4.6/34 = 0.135s to full speed, against the old 3.4/26 =
 * 0.131s, so the extra speed costs nothing in start latency. Left well short of
 * an instant snap on purpose — the stop still has a short slide, which is what
 * stops the knight looking like a cursor.
 */
const ACCEL = 34;

/** Above this speed the walk clip plays; below it the knight idles. */
const WALK_CLIP_THRESHOLD = 0.35;

const ISO = Math.SQRT1_2;

/** Current velocity, module-local (one player, one tavern). */
let vx = 0;
let vz = 0;
let animator: Animator | null = null;

/**
 * Snap the mesh so its texels land on whole render-target pixels. Unsnapped
 * sprites shimmer as they move — same trick, and same reason, as the dungeon's
 * `syncActorMesh`. But it has to be done against THIS scene's camera, and the
 * straight copy of the dungeon's version was not:
 *
 *  1. ZOOM. `Math.round(u * PPU) / PPU` quantises to 1/64 of a world unit, which
 *     is one render pixel only when `camera.zoom === 1`. The dungeon never sets
 *     zoom, so that holds there. The tavern runs at CAM_ZOOM_WIDE 0.78 and eases
 *     to CAM_ZOOM_FOCUS 0.92 whenever a station takes focus (core.ts), so a
 *     1/PPU step was 0.78 of a pixel — SUB-pixel, so it never aligned a texel to
 *     anything, and the rounding error it added was pure noise. The snap was
 *     doing none of the job it was there for and all of the damage.
 *  2. FRAME OF REFERENCE. The dungeon's camera has a dead-zone, so it is usually
 *     STILL and a world-space snap is also a screen-space snap. The tavern's
 *     camera eases toward its target every single frame with no dead-zone, so
 *     the world slid smoothly underneath a sprite that was hopping on a fixed
 *     world lattice — the sprite visibly juddered against the background even
 *     when the player was walking in a straight line. That is the "not smooth"
 *     in the report, and it is why raising WALK_SPEED alone would not have
 *     fixed it: a faster hop is a bigger hop.
 *
 * So: snap to a real render pixel (1/(PPU * zoom)), and snap RELATIVE TO THE
 * CAMERA rather than to the world origin. The sprite's offset from the camera
 * lands on whole pixels — which is the whole point, it is what keeps the texels
 * from shimmering — while the camera's own smooth pan carries the sprite along
 * with the scenery instead of fighting it.
 *
 * TRADEOFF, stated plainly: the sprite is no longer pinned to a fixed world
 * lattice, so relative to the *floor* it can shift by up to half a pixel as the
 * camera eases. That is invisible (the floor is a flat quantised texture and the
 * camera moves anyway); the judder it replaces was not. Falls back to the old
 * behaviour when there is no camera yet, which is the case during
 * `createTavernPlayer`.
 */
function syncMesh(p: TavernPlayer): void {
  const u = (p.x - p.z) * ISO;
  const v = (p.x + p.z) * ISO;
  const zoom = tavern.camera?.zoom ?? 1;
  // One render-target pixel, expressed in world units.
  const texel = 1 / (PPU * zoom);
  // The camera's aim point in the same (u, v) camera-aligned basis.
  const cu = (tavern.camX - tavern.camZ) * ISO;
  const cv = (tavern.camX + tavern.camZ) * ISO;
  const su = cu + Math.round((u - cu) / texel) * texel;
  const sv = cv + Math.round((v - cv) / texel) * texel;
  p.sprite.mesh.position.set((sv + su) * ISO, 0, (sv - su) * ISO);
}

/** Build the knight and drop them at the foot of the stair. */
export function createTavernPlayer(scene: THREE.Scene): TavernPlayer {
  const weaponId = activeWeapon().id;
  // Reuse the dungeon's cached atlas when it exists, so entering the tavern
  // doesn't re-run the (not cheap) paint + pixel-crush for art we already have.
  let sheet = dungeonState.playerSheets.get(weaponId);
  if (!sheet) {
    sheet = buildSpriteSheet(makeKnightPaints(weaponId));
    dungeonState.playerSheets.set(weaponId, sheet);
  }

  const sprite = createActorSprite(sheet, false);
  scene.add(sprite.mesh);
  animator = new Animator(sprite);

  const p: TavernPlayer = {
    x: SPAWN.x,
    z: SPAWN.z,
    facing: "N" as Facing, // arriving from the stair, looking into the room
    speed: 0,
    animT: 0,
    sprite,
  };
  vx = 0;
  vz = 0;
  animator.setFacing(p.facing);
  syncMesh(p);
  return p;
}

/**
 * One movement step.
 *
 * `frozen` is true while a station panel is open — the knight holds position and
 * drops to idle rather than sliding around behind the UI.
 */
export function updateTavernPlayer(dt: number, input: InputHandle, frozen: boolean): void {
  const p = tavern.player;
  if (!p || !animator) return;

  let tx = 0;
  let tz = 0;
  if (!frozen) {
    const a = input.axis();
    if (a.x !== 0 || a.z !== 0) {
      // The input axis is SCREEN-relative; under the isometric yaw, screen-up is
      // a world diagonal. Rotating here is what makes "W" walk away from the
      // camera instead of off to one side.
      //
      // This MUST go through the same `screenDirToWorld` the dungeon uses. It
      // used to hand-roll the rotation as `(a.x - a.z, a.x + a.z) * ISO`, which
      // is the correct basis turned exactly 90°: W walked screen-RIGHT, A
      // walked screen-UP, S screen-LEFT, D screen-DOWN. The dungeon was fine
      // because it always called the shared helper — the tavern was the only
      // place with a second copy of the maths, and a second copy is the whole
      // reason the two could disagree. `movement.test.ts` now pins them equal.
      const w = screenDirToWorld(a.x, a.z);
      const wx = w.x;
      const wz = w.z;
      const len = Math.hypot(wx, wz) || 1;
      const speed = WALK_SPEED * (input.sprintHeld() ? HURRY_MULT : 1);
      tx = (wx / len) * speed;
      tz = (wz / len) * speed;
    }
  }

  // Ramp toward the target rather than snapping, so starts and stops have weight.
  const dvx = tx - vx;
  const dvz = tz - vz;
  const dvLen = Math.hypot(dvx, dvz);
  if (dvLen > 0) {
    const step = Math.min(dvLen, ACCEL * dt);
    vx += (dvx / dvLen) * step;
    vz += (dvz / dvLen) * step;
  }

  if (vx !== 0 || vz !== 0) {
    const wantDx = vx * dt;
    const wantDz = vz * dt;
    const moved = moveInRoom(p.x, p.z, p.x + wantDx, p.z + wantDz);
    // Kill the velocity component we just got blocked on, or we keep pressing
    // into the counter and the walk cycle plays on the spot.
    //
    // This used to be `moved.x === p.x` — an exact float compare, which is only
    // ever true on the SECOND frame of contact (the first frame still travels
    // the last sliver of gap to the ejection boundary, so the positions differ).
    // It happened to work because moveInRoom's ejection is deterministic, but it
    // meant one frame of full velocity buried in the counter, and it would go
    // silently dead the moment any float drift crept into the eject position.
    //
    // Comparing ACHIEVED against INTENDED is the honest question, and it is
    // tolerant: keep less than a tenth of what we asked for on an axis and that
    // axis is blocked. Note this only ever zeroes the BLOCKED axis — the free
    // one keeps its velocity, which is what makes brushing along a counter a
    // slide at full speed rather than a stutter where you re-accelerate from
    // zero every frame.
    if (Math.abs(moved.x - p.x) < Math.abs(wantDx) * 0.1) vx = 0;
    if (Math.abs(moved.z - p.z) < Math.abs(wantDz) * 0.1) vz = 0;
    p.x = moved.x;
    p.z = moved.z;
  }

  p.speed = Math.hypot(vx, vz);
  p.animT += dt;

  if (p.speed > WALK_CLIP_THRESHOLD) {
    p.facing = facingFromVelocity(vx, vz, p.facing);
    animator.setFacing(p.facing);
    animator.play("walk");
    // Gait quickens with speed so hurrying reads without a separate run clip.
    animator.setRate(0.7 + (p.speed / WALK_SPEED) * 0.6);
  } else {
    animator.play("idle");
    animator.setRate(1);
  }
  animator.update(dt);
  syncMesh(p);
}

/** Drop the module-local animator/velocity when the scene closes. */
export function disposeTavernPlayer(): void {
  animator = null;
  vx = 0;
  vz = 0;
}

File: crates/pk-core/src/tavern/player.rs
//! The knight, in the tavern. Walk, face, stop at furniture — the movement
//! math of `legacy/src/scenes/tavern/player.ts::updateTavernPlayer`, extracted
//! as a pure fixed-signature step so the legacy fixture exporter and this port
//! drive ONE description of the walk (mirrored in
//! `legacy/src/scenes/tavern/player.ts::stepTavernMovement`).
//!
//! Deliberately NOT the dungeon's 1500-line controller: axis → velocity ramp →
//! rect collision → facing is the whole job.
//!
//! PORTS: `legacy/src/scenes/tavern/player.ts`

use super::layout::{move_in_room, PLAYER_RADIUS, SPAWN};
use crate::jsmath::js_hypot;
use crate::state::Facing;

/// Walk speed, world units/sec — the dungeon's 4.2 plus a margin paid to the
/// tavern's camera lean (see the legacy comment on WALK_SPEED).
pub const WALK_SPEED: f64 = 4.6;
/// Hold shift to cross the room quickly without it feeling like a sprint mechanic.
pub const HURRY_MULT: f64 = 1.5;
/// Acceleration/deceleration, units/sec².
pub const ACCEL: f64 = 34.0;
/// Above this speed the walk clip plays; below it the knight idles.
pub const WALK_CLIP_THRESHOLD: f64 = 0.35;

/// legacy engine/camera.ts: ground-plane direction of "screen up" and
/// "screen right", derived from the engine yaw (45°, `Math.PI / 4`).
const YAW: f64 = std::f64::consts::FRAC_PI_4;

fn screen_up_xz() -> (f64, f64) {
    (-libm::sin(YAW), -libm::cos(YAW))
}
fn screen_right_xz() -> (f64, f64) {
    (libm::cos(YAW), -libm::sin(YAW))
}

/// Screen-space axis → world ground direction (legacy `screenDirToWorld`).
/// +sz is screen-down (toward the camera).
pub fn screen_dir_to_world(sx: f64, sz: f64) -> (f64, f64) {
    let up = screen_up_xz();
    let right = screen_right_xz();
    (sx * right.0 - sz * up.0, sx * right.1 - sz * up.1)
}

/// World ground direction → screen-space axis (legacy `worldDirToScreen`).
pub fn world_dir_to_screen(wx: f64, wz: f64) -> (f64, f64) {
    let up = screen_up_xz();
    let right = screen_right_xz();
    (wx * right.0 + wz * right.1, -(wx * up.0 + wz * up.1))
}

/// legacy engine/render/animator.ts `facingFromVelocity` — hysteresis so
/// diagonal jitter doesn't flap the sheet; sign reversals stay instant.
pub fn facing_from_velocity(vx: f64, vz: f64, fallback: Facing) -> Facing {
    let ax = vx.abs();
    let az = vz.abs();
    if ax < 1e-4 && az < 1e-4 {
        return fallback;
    }
    let vertical = fallback == Facing::S || fallback == Facing::N;
    const MARGIN: f64 = 1.25;
    if vertical && az * MARGIN >= ax && az >= 1e-4 {
        return if vz > 0.0 { Facing::S } else { Facing::N };
    }
    if !vertical && ax * MARGIN >= az && ax >= 1e-4 {
        return if vx > 0.0 { Facing::E } else { Facing::W };
    }
    // The other axis clearly won (or the current one went dead): switch. Ties
    // break toward the vertical axis, which reads better.
    if az >= ax {
        if vz > 0.0 {
            Facing::S
        } else {
            Facing::N
        }
    } else if vx > 0.0 {
        Facing::E
    } else {
        Facing::W
    }
}

/// The tavern knight's pose + velocity — everything one movement step touches.
#[derive(Debug, Clone, PartialEq)]
pub struct TavernPose {
    pub x: f64,
    pub z: f64,
    pub vx: f64,
    pub vz: f64,
    pub facing: Facing,
    /// Current speed, for the walk-cycle rate.
    pub speed: f64,
    /// Animation clock.
    pub anim_t: f64,
}

impl TavernPose {
    /// At the foot of the stair, looking into the room (legacy createTavernPlayer).
    pub fn spawn() -> Self {
        Self {
            x: SPAWN.0,
            z: SPAWN.1,
            vx: 0.0,
            vz: 0.0,
            facing: Facing::N,
            speed: 0.0,
            anim_t: 0.0,
        }
    }
}

/// Per-step input, already sampled by the shell.
#[derive(Debug, Clone, Copy, Default)]
pub struct TavernInput {
    /// Screen-relative axis, -1..1 per component (keyboard is ±1).
    pub axis_x: f64,
    pub axis_z: f64,
    pub sprint: bool,
    /// True while a station panel is open OR a one-shot clip owns the knight —
    /// the target velocity is zero but the ramp/collide/speed still run.
    pub frozen: bool,
}

/// One movement step — the exact math of legacy `updateTavernPlayer` (minus
/// the sprite/mesh work). Operation order matters: target from the rotated
/// axis, ramp toward it, move with per-axis blocked-velocity kill, then speed
/// and facing.
pub fn step_tavern_movement(p: &mut TavernPose, input: &TavernInput, dt: f64) {
    let mut tx = 0.0;
    let mut tz = 0.0;
    if !input.frozen && (input.axis_x != 0.0 || input.axis_z != 0.0) {
        // The input axis is SCREEN-relative; under the isometric yaw, screen-up
        // is a world diagonal. MUST go through the same screen_dir_to_world the
        // dungeon uses — a second copy of the maths is how the two scenes ever
        // disagreed (movement.test.ts pins the directions).
        let (wx, wz) = screen_dir_to_world(input.axis_x, input.axis_z);
        let len = {
            let h = js_hypot(wx, wz);
            if h == 0.0 {
                1.0
            } else {
                h
            }
        };
        let speed = WALK_SPEED * if input.sprint { HURRY_MULT } else { 1.0 };
        tx = (wx / len) * speed;
        tz = (wz / len) * speed;
    }

    // Ramp toward the target rather than snapping, so starts and stops have weight.
    let dvx = tx - p.vx;
    let dvz = tz - p.vz;
    let dv_len = js_hypot(dvx, dvz);
    if dv_len > 0.0 {
        let step = dv_len.min(ACCEL * dt);
        p.vx += (dvx / dv_len) * step;
        p.vz += (dvz / dv_len) * step;
    }

    if p.vx != 0.0 || p.vz != 0.0 {
        let want_dx = p.vx * dt;
        let want_dz = p.vz * dt;
        let (mx, mz) = move_in_room(p.x, p.z, p.x + want_dx, p.z + want_dz, PLAYER_RADIUS);
        // Kill the velocity component we just got blocked on — ACHIEVED vs
        // INTENDED, tolerant: keep less than a tenth of what we asked for on an
        // axis and that axis is blocked. Only ever zeroes the BLOCKED axis.
        if (mx - p.x).abs() < want_dx.abs() * 0.1 {
            p.vx = 0.0;
        }
        if (mz - p.z).abs() < want_dz.abs() * 0.1 {
            p.vz = 0.0;
        }
        p.x = mx;
        p.z = mz;
    }

    p.speed = js_hypot(p.vx, p.vz);
    p.anim_t += dt;

    if p.speed > WALK_CLIP_THRESHOLD {
        p.facing = facing_from_velocity(p.vx, p.vz, p.facing);
    }
}

#[cfg(test)]
mod tests {
    //! Direction cases ported from `legacy/src/scenes/tavern/movement.test.ts`.
    use super::*;

    fn unit(v: (f64, f64)) -> (f64, f64) {
        let len = {
            let h = js_hypot(v.0, v.1);
            if h == 0.0 {
                1.0
            } else {
                h
            }
        };
        (v.0 / len, v.1 / len)
    }

    /// The rotation the tavern used to apply, kept so the bug stays pinned.
    fn the_old_broken_way(sx: f64, sz: f64) -> (f64, f64) {
        const ISO: f64 = std::f64::consts::FRAC_1_SQRT_2;
        ((sx - sz) * ISO, (sx + sz) * ISO)
    }

    /// WASD as the input layer reports it: +z is screen-DOWN, toward the camera.
    const KEYS: [(&str, (f64, f64)); 4] = [
        ("W", (0.0, -1.0)),
        ("S", (0.0, 1.0)),
        ("A", (-1.0, 0.0)),
        ("D", (1.0, 0.0)),
    ];

    fn close(a: f64, b: f64, digits: i32) -> bool {
        (a - b).abs() < 0.5 * 10f64.powi(-digits)
    }

    #[test]
    fn sends_each_key_back_to_the_screen_axis_it_was_pressed_on() {
        for (key, axis) in KEYS {
            let w = screen_dir_to_world(axis.0, axis.1);
            let back = unit(world_dir_to_screen(w.0, w.1));
            assert!(close(back.0, axis.0, 6), "{key} screen-x");
            assert!(close(back.1, axis.1, 6), "{key} screen-z");
        }
    }

    #[test]
    fn does_not_reproduce_the_90_degree_rotation_the_tavern_used_to_have() {
        // W is the clearest single case: under the old maths it walked screen-right.
        let good = unit(screen_dir_to_world(0.0, -1.0));
        let bad = unit(the_old_broken_way(0.0, -1.0));
        assert!(!close(good.0, bad.0, 3));

        // Rotating the correct vector by +90° in XZ, (x, z) -> (-z, x), lands
        // exactly on the old one.
        assert!(close(bad.0, -good.1, 6));
        assert!(close(bad.1, good.0, 6));
    }

    #[test]
    fn keeps_opposite_keys_opposite_and_perpendicular_keys_perpendicular() {
        let w = unit(screen_dir_to_world(0.0, -1.0));
        let s = unit(screen_dir_to_world(0.0, 1.0));
        let d = unit(screen_dir_to_world(1.0, 0.0));

        assert!(close(w.0 + s.0, 0.0, 6));
        assert!(close(w.1 + s.1, 0.0, 6));
        // Perpendicular ⇒ dot product zero.
        assert!(close(w.0 * d.0 + w.1 * d.1, 0.0, 6));
    }

    #[test]
    fn gives_all_eight_keyboard_directions_a_distinct_heading() {
        let combos = [
            (0.0, -1.0),
            (1.0, -1.0),
            (1.0, 0.0),
            (1.0, 1.0),
            (0.0, 1.0),
            (-1.0, 1.0),
            (-1.0, 0.0),
            (-1.0, -1.0),
        ];
        let headings: std::collections::HashSet<String> = combos
            .iter()
            .map(|&(x, z)| {
                let w = unit(screen_dir_to_world(x, z));
                format!("{:.4},{:.4}", w.0, w.1)
            })
            .collect();
        assert_eq!(headings.len(), 8);
    }

    // ── Step semantics (behavioural pins on the extracted step) ──

    const DT: f64 = 1.0 / 60.0;

    #[test]
    fn ramps_to_full_speed_and_stops_with_a_slide() {
        let mut p = TavernPose::spawn();
        let hold = TavernInput {
            axis_x: 0.0,
            axis_z: -1.0, // W: away from the camera, into the room
            ..Default::default()
        };
        // 4.6/34 ≈ 0.135s to full speed — well inside half a second.
        for _ in 0..30 {
            step_tavern_movement(&mut p, &hold, DT);
        }
        assert!((p.speed - WALK_SPEED).abs() < 1e-9, "at full walk speed");
        assert_eq!(p.facing, Facing::N, "W walks away from the camera");

        let release = TavernInput::default();
        step_tavern_movement(&mut p, &release, DT);
        assert!(p.speed < WALK_SPEED, "decelerating");
        assert!(p.speed > 0.0, "but not an instant stop");
        for _ in 0..30 {
            step_tavern_movement(&mut p, &release, DT);
        }
        assert_eq!(p.speed, 0.0, "at rest");
    }

    #[test]
    fn frozen_zeroes_the_target_but_still_decelerates_honestly() {
        let mut p = TavernPose::spawn();
        let hold = TavernInput {
            axis_z: -1.0,
            ..Default::default()
        };
        for _ in 0..30 {
            step_tavern_movement(&mut p, &hold, DT);
        }
        let frozen = TavernInput {
            axis_z: -1.0,
            frozen: true,
            ..Default::default()
        };
        step_tavern_movement(&mut p, &frozen, DT);
        assert!(p.speed < WALK_SPEED, "input ignored while frozen");
    }

    #[test]
    fn sprint_multiplies_the_walk_speed() {
        let mut p = TavernPose::spawn();
        let hurry = TavernInput {
            axis_z: -1.0,
            sprint: true,
            ..Default::default()
        };
        for _ in 0..60 {
            step_tavern_movement(&mut p, &hurry, DT);
        }
        assert!((p.speed - WALK_SPEED * HURRY_MULT).abs() < 1e-9);
    }

    #[test]
    fn brushing_a_counter_keeps_the_free_axis_at_speed() {
        // Stand just south of the central table, push diagonally into it:
        // the z axis blocks, the x axis keeps its velocity (the slide).
        let table = &super::super::layout::OBSTACLES[0];
        let mut p = TavernPose::spawn();
        p.x = table.x - 0.6;
        p.z = table.z + table.d / 2.0 + PLAYER_RADIUS + 0.01;
        // Screen-axis (1, 0) is world (+,-) — east and into the table's face.
        let push = TavernInput {
            axis_x: 1.0,
            ..Default::default()
        };
        for _ in 0..20 {
            step_tavern_movement(&mut p, &push, DT);
        }
        assert_eq!(p.vz, 0.0, "blocked axis killed");
        assert!(p.vx > 0.0, "free axis still moving");
        assert!(p.x > table.x - 0.6, "made progress along the counter");
    }
}
